leaf-server/minecraft-patches/features/0277-optimize-checkInsideBlocks-calls.patch

```diff
@@ -0,0 +1,229 @@
+
+License: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)
+
+diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
+index 086525c31bd8167baac79a7f41e3e4d1ca783136..e05d1a4b4126c8cddcf9adbf83cc4b53c5b028d5 100644
+--- a/net/minecraft/world/entity/Entity.java
++++ b/net/minecraft/world/entity/Entity.java
+@@ -1733,8 +1733,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
+         }
+     }
+ 
++    // Leaf start - optimize checkInsideBlocks calls
+     private void checkInsideBlocks(Vec3 vec3, Vec3 vec31, InsideBlockEffectApplier.StepBasedCollector stepBasedCollector, LongSet set) {
+-        AABB aabb = this.makeBoundingBox(vec31).deflate(1.0E-5F);
++        final Level level = this.level();
++        AABB aabb = this.getBoundingBox().move(vec31.subtract(this.position())).deflate(1.0E-5F);
++        final net.minecraft.world.level.chunk.ChunkAccess[] cachedChunk = { null };
++        final long[] cachedChunkPos = { Long.MIN_VALUE };
++
+         BlockGetter.forEachBlockIntersectedBetween(
+             vec3,
+             vec31,
+@@ -1743,35 +1748,51 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
+                 if (!this.isAlive()) {
+                     return false;
+                 } else {
+-                    BlockState blockState = this.level().getBlockState(pos);
++                    final int chunkX = pos.getX() >> 4;
++                    final int chunkZ = pos.getZ() >> 4;
++                    final long currentChunkPos = ChunkPos.asLong(chunkX, chunkZ);
++                    BlockState blockState;
++
++                    if (cachedChunkPos[0] != currentChunkPos) {cachedChunk[0] = level.getChunkIfLoaded(chunkX, chunkZ);cachedChunkPos[0] = currentChunkPos;}
++
++                    if (cachedChunk[0] != null) {blockState = cachedChunk[0].getBlockState(pos);
++                    } else {
++                        blockState = level.getBlockStateIfLoaded(pos);
++                        if (blockState == null) {blockState = Blocks.AIR.defaultBlockState();}
++                    }
++
++                    final long posLong = pos.asLong();
+                     if (blockState.isAir()) {
+                         this.debugBlockIntersection(pos, false, false);
+                         return true;
+-                    } else if (!set.add(pos.asLong())) {
++                    } else if (!set.add(posLong)) {
+                         return true;
+                     } else {
+-                        VoxelShape entityInsideCollisionShape = blockState.getEntityInsideCollisionShape(this.level(), pos, this);
+-                        boolean flag = entityInsideCollisionShape == Shapes.block()
+-                            || this.collidedWithShapeMovingFrom(vec3, vec31, entityInsideCollisionShape.move(new Vec3(pos)).toAabbs());
++                        final Vec3 blockVec = new Vec3(pos);
```

Comment:
Putting this in a variable seems unnecessary? And it's slightly slower this way since in its original position it might not even be executed sometimes.